docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+be added which extends this service and bundles the new core services, ex:
+
+```go
+package v2
+
+import "cosmossdk.io/core/appmodule"
+
+type Service interface {
+	appmodule.Service
+	SomeNewService
+	AnotherNewService
+}
+```
+
+### Core `Handler` Struct
+
+
+Modules will provide their core services to runtime module via the `Handler` struct, defined in the `cosmossdk.io/core/appmodule`
+package, instead of the existing
+`AppModule` interface:
+```go
+package appmodule
+
+type Handler struct {
+    // Services are the Msg and Query services for the module. Msg services
+    // must be annotated with the option cosmos.msg.v1.service = true.
+    Services []ServiceImpl
+	
+	DefaultGenesis func(GenesisTarget)
+    ValidateGenesis func(GenesisSource) error
+    InitGenesis func(context.Context, GenesisSource) error
+    ExportGenesis func(context.Context, GenesisTarget)
+
+    BeginBlocker func(context.Context) error
+	EndBlocker func(context.Context) error
+	
+	EventListeners []EventListener
+	
+	UpgradeHandlers []UpgradeHandler
+}
+```
+
+A struct as opposed to an interface has been chosen for the following reasons:
+* it is always possible to add new fields to a struct without breaking backwards compatibility
+* it is not necessary to populate all fields in a struct, whereas interface methods must always be implemented
+* compared to extension interfaces, struct fields are more explicit
+
+Helper methods will be added to the `Handler` struct to simplify construction of more complex fields.
+
+#### `MsgServer` and `QueryServer` registration
+
+The `Handler` struct type will implement `grpc.ServiceRegistrar` interface which will allow for registering
+`MsgServer` and `QueryServer` implementations directly with the `Handler` struct like this:
+```go
+    h := &appmodule.Handler{}
+	types.RegisterMsgServer(h, keeper.NewMsgServerImpl(am.keeper))
+	types.RegisterQueryServer(h, am.keeper)
+```
+
+Service registrations will be stored in the `Handler.Services` field using the `ServiceImpl` struct:
+```go
+type ServiceImpl struct {
+	Desc *grpc.ServiceDesc
+	Impl interface{}
+}
+```
+
+Because of the `cosmos.msg.v1.service` protobuf option, required for `Msg` services, the same handler struct can be
+used to register both `Msg` and query services.
+
+#### Genesis
+
+The genesis `Handler` functions - `DefaultGenesis`, `ValidateGenesis`, `InitGenesis` and `ExportGenesis` - are specified
+against the `GenesisSource` and `GenesisTarget` interfaces which will abstract over genesis sources which may be single
+JSON files or collections of JSON objects that can be efficiently streamed. They will also include helper functions for
+interacting with genesis data represented by single `proto.Message` types.
+
+```go
+type GenesisSource interface {
+	ReadMessage(proto.Message) error
+	OpenReader(field string) (io.ReadCloser, error)
+	ReadRawJSON() (json.RawMessage, error)
```

Comment:
we should opt to use one and abstract the other, this is more boiler plate for users

Reply:
Why is it more boilerplate? The user doesn't implement this interface, they just use it how they like.